### PR TITLE
Add xades support

### PR DESCRIPTION
**Pull request by pokoli**
Hi, 

Here is the implementation of [#90](https://github.com/XML-Security/signxml/issues/90)
For now we just include basic properties of xades but it works for simple implementations. 

I will like to have some feedback of the maintainers to now what is required to include this feature as part of a new release. Do we need full xades support or we can start which just basic functionalities? 

P.S: There are a lot of old commits here, do you think it makes sense to merge them into a single one to make the review easier?

### CODE REVIEW

**Comment by pokoli** [comment]
Note that on last commit I use the default parameter of namedtuple which is only available on python 3.7. 
I've created #200 which removes support for Python 3.6, this is a requirement to allow merging this PR.

File: signxml/__init__.py [hunk 1]
```diff
@@ -370,9 +370,9 @@ def sign(
             self.id_attributes = (id_attribute,)
 
         if isinstance(cert, (str, bytes)):
-            cert_chain = list(iterate_pem(cert))
+            self.cert_chain = list(iterate_pem(cert))
```
Comment (kislyuk) [inline signxml/__init__.py:373]: Please do not attach any signature-specific information to the signer/verifier instance. The instance is stateless.

File: signxml/__init__.py [hunk 2]
```diff
@@ -526,6 +529,27 @@ def _unpack(self, data, reference_uris):
         return sig_root, doc_root, c14n_inputs, reference_uris
 
     def _build_sig(self, sig_root, reference_uris, c14n_inputs, sig_insp, payload_insp):
+        """
+        :param reference_uris: the references to include o buld.
```
Comment (kislyuk) [inline signxml/__init__.py:533]:
```suggestion
        :param reference_uris: the references to include or build.
```

File: signxml/__init__.py [hunk 3]
```diff
@@ -526,6 +529,27 @@ def _unpack(self, data, reference_uris):
...
+        - elements: if the reference_uri is an element, this will be added
+          without modifications
+        - dict: if the reference_uri is a dict, the reference element will be
+          created using the attributes in the dict
+          i.e: {"URI": URI, "Type": Type, "Id": ID}
+        - string: if refence_uri is a string, this will be interpreted as the
+          attribute `URI`
+        """
+
+        def _check_brothers(element1, element2):
```
Comment (kislyuk) [inline signxml/__init__.py:544]: This doesn't seem to be used anywhere?

File: signxml/xades/README.rst [hunk 4]
```diff
@@ -0,0 +1,49 @@
+#############
+XAdES by ETSI
+#############
+
+`W3C Note: XML Advanced Electronic Signatures (XAdES) <https://www.w3.org/TR/XAdES>`_ is outdated and has aparently been superseeded by standard development by `ETSI <https://www.etsi.org/>`_.
```
Comment (kislyuk) [inline signxml/xades/README.rst:5]:
```suggestion
`W3C Note: XML Advanced Electronic Signatures (XAdES) <https://www.w3.org/TR/XAdES>`_ is outdated and has apparently been superseded by standard development by `ETSI <https://www.etsi.org/>`_.
```

File: signxml/xades/__init__.py [hunk 5]
```diff
@@ -0,0 +1,681 @@
...
+        self.black_list = black_list
+        self.options_struct = options
+        self.refs = []
+        super().__init__()
+
+    def _unpack(self, data, reference_uris):
+        sig_root, doc_root, c14n_inputs, reference_uris = super()._unpack(data, reference_uris)
+        self._set_default_options()
+        reference_uris.append(DS.Object(self._generate_xades()))
+        # reference_uris.extend(self.refs)
```
Comment (kislyuk) [inline signxml/xades/__init__.py:200]:
```suggestion
```

File: signxml/xades/__init__.py [hunk 6]
```diff
@@ -0,0 +1,681 @@
...
+                encoding of one DER-encoded instance of type IssuerSerial type
+                defined in IETF RFC 5035 [17].
+                """
+                raise NotImplementedError(
+                    "Please make a PR if you know how to obtain in python a "
+                    "'DER-encoded instance of type IssuerSerial type defined in "
+                    "IETF RFC 5035'"
+                )
+                serial_element = XADES.IssuerSerialV2(
+                    # TODO implement, wtf?
```
Comment (kislyuk) [inline signxml/xades/__init__.py:305]:
```suggestion
                    # TODO: implement
```

File: signxml/xades/__init__.py [hunk 7]
```diff
@@ -0,0 +1,681 @@
...
+            • attribute certificates (issued, in consequence, by Attribute
+              Authorities) in different syntax than the one specified in
+              Recommendation ITU-T X.509 [4], within the
+              OtherAttributeCertificate element. The definition of specific
+              OtherAttributeCertificate is outside of the scope of the present
+              document
+        """
+        if self.options_struct.CertifiedRoles and self.xades_legacy:
+            NotImplementedError(
+                "Legay certified roles wired as objects encoded in " "EncapsulatedPKIDataType are not implemented."
```
Comment (kislyuk) [inline signxml/xades/__init__.py:433]:
```suggestion
                "Legacy certified roles wired as objects encoded in " "EncapsulatedPKIDataType are not implemented."
```

File: signxml/xades/__init__.py [hunk 8]
```diff
@@ -0,0 +1,681 @@
...
+        if not sr_elements:
+            pass
+        else:
+            sr_elements = self._clean_elements_from_black_list(sr_elements)
+            if self.xades_legacy:
+                elements.append(XADES.SignerRole(*sr_elements))  # deprecated (legacy)
+            else:
+                elements.append(XADES.SignerRoleV2(*sr_elements))
+
+        # any ##other
```
Comment (kislyuk) [inline signxml/xades/__init__.py:482]:
```suggestion
```

File: signxml/xades/__init__.py [hunk 9]
```diff
@@ -0,0 +1,681 @@
...
+        elements: List[ElementMaker] = []
+        return elements
+
+        # TODO: Find a way to complete such values
+        elements.append(XADES.DataObjectFormat())
+        elements.append(XADES.CommitmentTypeIndication())
+        elements.append(XADES.AllDataObjectsTimeStamp())
+        elements.append(XADES.IndividualDataObjectsTimeStamp())
+
+        # any ##other
```
Comment (kislyuk) [inline signxml/xades/__init__.py:496]:
```suggestion
```

File: signxml/xades/__init__.py [hunk 10]
```diff
@@ -0,0 +1,681 @@
...
+        else:
+            elements.append(XADES141.RefsOnlyTimeStampV2())
+
+        elements.append(XADES.CertificateValues())
+        elements.append(XADES.RevocationValues())
+        elements.append(XADES.AttrAuthoritiesCertValues())
+        elements.append(XADES.AttributeRevocationValues())
+        elements.append(XADES.ArchiveTimeStamp())
+
+        # any ##other
```
Comment (kislyuk) [inline signxml/xades/__init__.py:541]:
```suggestion
```

**Review by kislyuk — CHANGES_REQUESTED** [review]
This is a great start, thanks.

I have added a few comments inline. Please make sure there are no stray comments and unused code left behind. All methods should be used by other methods or documented as part of the public API.

Can you please add an interoperability ("interop") unit test case that verifies third party documents ("vectors") signed using xades? It looks like the following repos have a few test files:
* https://github.com/esig/dss/tree/master/dss-asic-xades/src/test/resources
* https://github.com/Ahmet-Kaplan/xades4j.interop/tree/master/src/test/resources/verification

**Comment by kislyuk** [comment]
Thanks everyone for your work on this PR. XAdES support was added in v3.0.0 using a different code structure. Please test v3.0.0 and let us know whether it suits your needs.

**Comment by ovnicraft** [comment]
@kislyuk great news! any code recommended to read ? I would like to improve documentation and add some examples about XaDES

**Comment by ovnicraft** [comment]
Found it http://xml-security.github.io/signxml/#xades-signatures !

**Pull request closed without merging**